// humane-indexer/lib/schemas/mapping_types.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
// jscs:disable requireCamelCaseOrUpperCaseIdentifiers
const $Keyword = exports.$Keyword = {
    type: 'string',
    analyzer: 'keyword_analyzer'
};

const $VernacularKeyword = exports.$VernacularKeyword = {
    type: 'string',
    analyzer: 'keyword_analyzer'
};

const $Long = exports.$Long = {
    type: 'long'
};

const $NotIndexedLong = exports.$NotIndexedLong = {
    type: 'long',
    index: 'no',
    include_in_all: false
};

const $Double = exports.$Double = {
    type: 'double'
};

const $NotIndexedDouble = exports.$NotIndexedDouble = {
    type: 'double',
    index: 'no',
    include_in_all: false
};

const $Boolean = exports.$Boolean = {
    type: 'boolean'
};

const $NotIndexedBoolean = exports.$NotIndexedBoolean = {
    type: 'boolean',
    index: 'no',
    include_in_all: false
};

const $Date = exports.$Date = {
    type: 'date',
    format: 'yyyy-MM-dd HH:mm:ss||epoch_millis'
};

const $NotIndexedText = exports.$NotIndexedText = {
    type: 'string',
    index: 'no',
    include_in_all: false
};

const $IdentityText = exports.$IdentityText = {
    type: 'string',
    index: 'not_analyzed'
};

const $Text = exports.$Text = {
    type: 'string',
    analyzer: 'standard_index_analyzer',
    search_analyzer: 'standard_search_analyzer',
    fields: {
        raw: $Keyword,
        edgeGram: {
            type: 'string',
            analyzer: 'edgeGram_index_analyzer',
            search_analyzer: 'standard_search_analyzer'
        },
        nGram: {
            type: 'string',
            analyzer: 'nGram_index_analyzer',
            search_analyzer: 'standard_search_analyzer'
        },
        phonetic_bm: {
            type: 'string',
            analyzer: 'phonetic_bm_index_analyzer',
            search_analyzer: 'phonetic_bm_search_analyzer'
        },
        phonetic_edgeGram_bm: {
            type: 'string',
            analyzer: 'phonetic_edgeGram_bm_index_analyzer',
            search_analyzer: 'phonetic_bm_search_analyzer'
        },
        phonetic_soundex: {
            type: 'string',
            analyzer: 'phonetic_soundex_index_analyzer',
            search_analyzer: 'phonetic_soundex_search_analyzer'
        },
        phonetic_edgeGram_soundex: {
            type: 'string',
            analyzer: 'phonetic_edgeGram_soundex_index_analyzer',
            search_analyzer: 'phonetic_soundex_search_analyzer'
        },
        phonetic_dm: {
            type: 'string',
            analyzer: 'phonetic_dm_index_analyzer',
            search_analyzer: 'phonetic_dm_search_analyzer'
        },
        phonetic_edgeGram_dm: {
            type: 'string',
            analyzer: 'phonetic_edgeGram_dm_index_analyzer',
            search_analyzer: 'phonetic_dm_search_analyzer'
        }
    }
};

const $VernacularText = exports.$VernacularText = {
    type: 'string',
    analyzer: 'standard_index_analyzer',
    search_analyzer: 'standard_search_analyzer',
    fields: {
        raw: $Keyword,
        edgeGram: {
            type: 'string',
            analyzer: 'edgeGram_index_analyzer',
            search_analyzer: 'standard_search_analyzer'
        }
    }
};
